From f2cfcbc44ed8600194ef0bf1ce343d65b7bbd9d4 Mon Sep 17 00:00:00 2001
From: Keunhong Park <keunhong@codeaurora.org>
Date: Wed, 11 Jul 2012 15:51:03 -0600
Subject: [PATCH] Add support for binding CacheRemovalListenerFactory

This can be used to dynamically bind removal listeners for use in
multi-master support.

Change-Id: I9c58740c54c89e5bd75accc64f03f1fdddc8d581
---
 .../server/cache/h2/CacheRemovalListener.java      |   41 +++++++++++++
 .../cache/h2/CacheRemovalListenerFactory.java      |   36 ++++++++++++
 .../server/cache/h2/DefaultCacheFactory.java       |   12 ++++-
 .../server/cache/h2/MasterRemovalListener.java     |   61 ++++++++++++++++++++
 .../cache/h2/MasterRemovalListenerFactory.java     |   58 +++++++++++++++++++
 gerrit-plugin-api/pom.xml                          |    8 ---
 6 files changed, 207 insertions(+), 9 deletions(-)
 create mode 100644 gerrit-cache-h2/src/main/java/com/google/gerrit/server/cache/h2/CacheRemovalListener.java
 create mode 100644 gerrit-cache-h2/src/main/java/com/google/gerrit/server/cache/h2/CacheRemovalListenerFactory.java
 create mode 100644 gerrit-cache-h2/src/main/java/com/google/gerrit/server/cache/h2/MasterRemovalListener.java
 create mode 100644 gerrit-cache-h2/src/main/java/com/google/gerrit/server/cache/h2/MasterRemovalListenerFactory.java

diff --git a/gerrit-cache-h2/src/main/java/com/google/gerrit/server/cache/h2/CacheRemovalListener.java b/gerrit-cache-h2/src/main/java/com/google/gerrit/server/cache/h2/CacheRemovalListener.java
new file mode 100644
index 0000000..7d67b97
--- /dev/null
+++ b/gerrit-cache-h2/src/main/java/com/google/gerrit/server/cache/h2/CacheRemovalListener.java
@@ -0,0 +1,41 @@
+// Copyright (c) 2012, Code Aurora Forum. All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+//     * Redistributions of source code must retain the above copyright
+//       notice, this list of conditions and the following disclaimer.
+//     * Redistributions in binary form must reproduce the above
+//       copyright notice, this list of conditions and the following
+//       disclaimer in the documentation and/or other materials provided
+//       with the distribution.
+//     * Neither the name of Code Aurora Forum, Inc. nor the names of its
+//       contributors may be used to endorse or promote products derived
+//       from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED "AS IS" AND ANY EXPRESS OR IMPLIED
+// WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS
+// BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
+// BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE
+// OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN
+// IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+package com.google.gerrit.server.cache.h2;
+
+import com.google.common.cache.RemovalListener;
+
+public abstract class CacheRemovalListener<K, V> implements
+    RemovalListener<K, V> {
+  protected String pluginName;
+  protected String cacheName;
+
+  public CacheRemovalListener(String pluginName, String cacheName) {
+    this.pluginName = (pluginName == null) ? "gerrit" : pluginName;
+    this.cacheName = cacheName;
+  }
+}
diff --git a/gerrit-cache-h2/src/main/java/com/google/gerrit/server/cache/h2/CacheRemovalListenerFactory.java b/gerrit-cache-h2/src/main/java/com/google/gerrit/server/cache/h2/CacheRemovalListenerFactory.java
new file mode 100644
index 0000000..1442409
--- /dev/null
+++ b/gerrit-cache-h2/src/main/java/com/google/gerrit/server/cache/h2/CacheRemovalListenerFactory.java
@@ -0,0 +1,36 @@
+// Copyright (c) 2012, Code Aurora Forum. All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+//     * Redistributions of source code must retain the above copyright
+//       notice, this list of conditions and the following disclaimer.
+//     * Redistributions in binary form must reproduce the above
+//       copyright notice, this list of conditions and the following
+//       disclaimer in the documentation and/or other materials provided
+//       with the distribution.
+//     * Neither the name of Code Aurora Forum, Inc. nor the names of its
+//       contributors may be used to endorse or promote products derived
+//       from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED "AS IS" AND ANY EXPRESS OR IMPLIED
+// WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS
+// BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
+// BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE
+// OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN
+// IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+package com.google.gerrit.server.cache.h2;
+
+import com.google.gerrit.extensions.annotations.ExtensionPoint;
+
+@ExtensionPoint
+public interface CacheRemovalListenerFactory {
+  public <K, V> CacheRemovalListener<K, V> get(String pluginName,
+      String cacheName);
+}
diff --git a/gerrit-cache-h2/src/main/java/com/google/gerrit/server/cache/h2/DefaultCacheFactory.java b/gerrit-cache-h2/src/main/java/com/google/gerrit/server/cache/h2/DefaultCacheFactory.java
index 8bb0709..f2dc627 100644
--- a/gerrit-cache-h2/src/main/java/com/google/gerrit/server/cache/h2/DefaultCacheFactory.java
+++ b/gerrit-cache-h2/src/main/java/com/google/gerrit/server/cache/h2/DefaultCacheFactory.java
@@ -20,6 +20,7 @@ import com.google.common.cache.CacheBuilder;
 import com.google.common.cache.CacheLoader;
 import com.google.common.cache.LoadingCache;
 import com.google.common.cache.Weigher;
+import com.google.gerrit.extensions.registration.DynamicSet;
 import com.google.gerrit.lifecycle.LifecycleModule;
 import com.google.gerrit.server.cache.CacheBinding;
 import com.google.gerrit.server.cache.MemoryCacheFactory;
@@ -28,6 +29,7 @@ import com.google.gerrit.server.cache.h2.H2CacheImpl.ValueHolder;
 import com.google.gerrit.server.config.ConfigUtil;
 import com.google.gerrit.server.config.GerritServerConfig;
 import com.google.inject.Inject;
+import com.google.inject.Scopes;
 
 import org.eclipse.jgit.lib.Config;
 
@@ -37,6 +39,9 @@ public class DefaultCacheFactory implements MemoryCacheFactory {
   public static class Module extends LifecycleModule {
     @Override
     protected void configure() {
+      DynamicSet.setOf(binder(), CacheRemovalListenerFactory.class);
+
+      bind(MasterRemovalListenerFactory.class).in(Scopes.SINGLETON);
       bind(DefaultCacheFactory.class);
       bind(MemoryCacheFactory.class).to(DefaultCacheFactory.class);
       bind(PersistentCacheFactory.class).to(H2CacheFactory.class);
@@ -45,10 +50,13 @@ public class DefaultCacheFactory implements MemoryCacheFactory {
   }
 
   private final Config cfg;
+  private MasterRemovalListenerFactory masterRemovalListenerFactory;
 
   @Inject
-  public DefaultCacheFactory(@GerritServerConfig Config config) {
+  public DefaultCacheFactory(@GerritServerConfig Config config,
+      MasterRemovalListenerFactory masterRemovalListenerFactory) {
     this.cfg = config;
+    this.masterRemovalListenerFactory = masterRemovalListenerFactory;
   }
 
   @Override
@@ -73,6 +81,8 @@ public class DefaultCacheFactory implements MemoryCacheFactory {
         "cache", def.name(), "memoryLimit",
         def.maximumWeight()));
 
+    builder.removalListener(masterRemovalListenerFactory.get(def.name()));
+
     Weigher<K, V> weigher = def.weigher();
     if (weigher != null && unwrapValueHolder) {
       final Weigher<K, V> impl = weigher;
diff --git a/gerrit-cache-h2/src/main/java/com/google/gerrit/server/cache/h2/MasterRemovalListener.java b/gerrit-cache-h2/src/main/java/com/google/gerrit/server/cache/h2/MasterRemovalListener.java
new file mode 100644
index 0000000..8e7050e
--- /dev/null
+++ b/gerrit-cache-h2/src/main/java/com/google/gerrit/server/cache/h2/MasterRemovalListener.java
@@ -0,0 +1,61 @@
+// Copyright (c) 2012, Code Aurora Forum. All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+//     * Redistributions of source code must retain the above copyright
+//       notice, this list of conditions and the following disclaimer.
+//     * Redistributions in binary form must reproduce the above
+//       copyright notice, this list of conditions and the following
+//       disclaimer in the documentation and/or other materials provided
+//       with the distribution.
+//     * Neither the name of Code Aurora Forum, Inc. nor the names of its
+//       contributors may be used to endorse or promote products derived
+//       from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED "AS IS" AND ANY EXPRESS OR IMPLIED
+// WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS
+// BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
+// BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE
+// OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN
+// IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+package com.google.gerrit.server.cache.h2;
+
+import com.google.common.cache.RemovalListener;
+import com.google.common.cache.RemovalNotification;
+import com.google.gerrit.extensions.registration.DynamicSet;
+
+/**
+ * This listener dispatches removal events to all other RemovalListeners
+ * attached using through the DynamicSet API.
+ *
+ * @param <K>
+ * @param <V>
+ */
+public class MasterRemovalListener<K, V> implements RemovalListener<K, V> {
+  private DynamicSet<CacheRemovalListenerFactory> listenerFactories;
+  private String pluginName;
+  private String cacheName;
+
+  public MasterRemovalListener(
+      DynamicSet<CacheRemovalListenerFactory> listenerFactories,
+      String pluginName, String cacheName) {
+    this.listenerFactories = listenerFactories;
+    this.pluginName = pluginName;
+    this.cacheName = cacheName;
+  }
+
+  @Override
+  public void onRemoval(RemovalNotification<K, V> notification) {
+    for (CacheRemovalListenerFactory factory : listenerFactories) {
+      RemovalListener<K, V> listener = factory.get(pluginName, cacheName);
+      listener.onRemoval(notification);
+    }
+  }
+}
diff --git a/gerrit-cache-h2/src/main/java/com/google/gerrit/server/cache/h2/MasterRemovalListenerFactory.java b/gerrit-cache-h2/src/main/java/com/google/gerrit/server/cache/h2/MasterRemovalListenerFactory.java
new file mode 100644
index 0000000..d408f5d
--- /dev/null
+++ b/gerrit-cache-h2/src/main/java/com/google/gerrit/server/cache/h2/MasterRemovalListenerFactory.java
@@ -0,0 +1,58 @@
+// Copyright (c) 2012, Code Aurora Forum. All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+//     * Redistributions of source code must retain the above copyright
+//       notice, this list of conditions and the following disclaimer.
+//     * Redistributions in binary form must reproduce the above
+//       copyright notice, this list of conditions and the following
+//       disclaimer in the documentation and/or other materials provided
+//       with the distribution.
+//     * Neither the name of Code Aurora Forum, Inc. nor the names of its
+//       contributors may be used to endorse or promote products derived
+//       from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED "AS IS" AND ANY EXPRESS OR IMPLIED
+// WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS
+// BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
+// BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE
+// OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN
+// IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+package com.google.gerrit.server.cache.h2;
+
+import com.google.gerrit.extensions.annotations.PluginName;
+import com.google.gerrit.extensions.registration.DynamicSet;
+import com.google.inject.Inject;
+
+/**
+ * Creates instances of MasterRemovalListers. This is required since
+ * RemovalListers take generic arguments from the creating method.
+ *
+ */
+public class MasterRemovalListenerFactory {
+  private DynamicSet<CacheRemovalListenerFactory> listenerFactories;
+  private String pluginName;
+
+  @Inject
+  public MasterRemovalListenerFactory(
+      DynamicSet<CacheRemovalListenerFactory> listenerFactories) {
+    this.listenerFactories = listenerFactories;
+  }
+
+  @Inject(optional = true)
+  void setPluginName(@PluginName String pluginName) {
+    this.pluginName = pluginName;
+  }
+
+  public <K, V> MasterRemovalListener<K, V> get(String cacheName) {
+    return new MasterRemovalListener<K, V>(listenerFactories, pluginName,
+        cacheName);
+  }
+}
diff --git a/gerrit-plugin-api/pom.xml b/gerrit-plugin-api/pom.xml
index 84f6f7b..0e78977 100644
--- a/gerrit-plugin-api/pom.xml
+++ b/gerrit-plugin-api/pom.xml
@@ -87,14 +87,6 @@ limitations under the License.
               <exclude>oro:oro</exclude>
             </excludes>
           </artifactSet>
-          <filters>
-            <filter>
-              <artifact>com.google.gerrit:gerrit-server</artifact>
-              <excludes>
-                <exclude>gerrit/**</exclude>
-              </excludes>
-            </filter>
-          </filters>
         </configuration>
         <executions>
           <execution>
-- 
1.7.8.3

